### ISSUE TITLE
False positive used-before-assignment for variable annotations when using TYPE_CHECKING

### ISSUE THREAD

**Issue by mew1033** [PyCQA/pylint#7609]
### Bug description

I know this has been looked at before, but I just ran into it again with pylint 2.15.4. If an import is guarded behind TYPE_CHECKING, I'm getting an `used-before-assignment` error when using that import in a type definition.

### Command used

```shell
pylint file.py
```

### Pylint output

```shell
file.py:7:19: E0601: Using variable 'PublishBatchRequestEntryTypeDef' before assignment (used-before-assignment)
```

### Expected behavior

It shouldn't error

### Pylint version

```shell
pylint 2.15.4
astroid 2.12.11
Python 3.9.14 (main, Sep  7 2022, 23:43:29) 
[GCC 9.4.0]
```

**Comment by mbyrnepr2** [comment]
Thanks for reporting this @mew1033. Can you please provide a minimal reproducible example to help us dig into this one?

**Comment by mew1033** [comment]
@mbyrnepr2 So sorry, I got the example ready and then completely forgot to add it to the issue. Here's a snippet that shows the issue. If you remove the `if TYPE_CHECKING:` line, the issues goes away.

```python
from typing import TYPE_CHECKING
import boto3

if TYPE_CHECKING:
    from mypy_boto3_sns.type_defs import PublishBatchRequestEntryTypeDef


def publish_to_sns():
    """
    Create SNS messages and publish them
    """
    sns_client = boto3.client("sns")
    messages: list[PublishBatchRequestEntryTypeDef] = []
```

**Comment by mbyrnepr2** [comment]
No worries - thank you!

**Comment by jacobtylerwalls** [comment]
Thanks! Per [PEP 526](https://peps.python.org/pep-0526/#runtime-effects-of-type-annotations), there is a difference in the runtime effect of variable annotations versus function annotations. Given that, the requirement in 3159b1744caee14068b4d875bb5adaa5ffbebd44 that both branches of an `if TYPE_CHECKING:` case define the name is too strict for variable annotations.

**Comment by jacobtylerwalls** [comment]
Some helpful test cases given in #7773
